### PR TITLE
Fixed typing

### PR DESCRIPTION
**Pull request by ssbarnea**
**Changelog-friendly one-liner**: Fixed typing

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

pip 21.0 has internal changes that break at least our typing and we will need significant changes to support 21.x for typing. For the moment we ceil pip version for mypy and assure running mypy does get us consistentresults inside and outside pre-commit.

Fixes type change introduced by click 8

A previous version of this patch attempted to pin down pip version in setup.cfg but I soon found that piplatest tox env
started to fail as `in-tree-build` feature was unknown to it. I suppose we will have to bump minimal pip version required
to >=21 as soon we upgrade type hints .

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -93,5 +93,9 @@ warn_return_any = true
 warn_unused_configs = true
 warn_unused_ignores = true
 
+# Avoid error: Duplicate module named 'setup'
+# https://github.com/python/mypy/issues/4008
+exclude = ^tests/test_data/
```
Comment (webknjaz) [inline setup.cfg:98]: I think we should update the invocation instead.

Comment (ssbarnea) [reply setup.cfg:98]: I will look into addressing this in a follow-up as this fix is unrelated to current change (the exclude was already there).  I am sure we will have more mypy work done soon, but now I need to unblock the CI/CD gates.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -25,7 +25,14 @@ repos:
       - id: mypy
         # Avoid error: Duplicate module named 'setup'
         # https://github.com/python/mypy/issues/4008
+        # Keep exclude in sync with mypy own excludes
         exclude: ^tests/test_data/
+        entry: mypy
```
Comment (asottile) [inline .pre-commit-config.yaml:30]: this line is the default and doesn't do anything

**Comment by ssbarnea** [comment]
I think we should ignore the coverage issue as covering for both versions of click 7 and 8 would make our CI pipelines too complex. It is just one line not covered. It would be much easier to just bump minimal version of click required than covering for both versions.

File: piptools/scripts/compile.py [hunk 3]
```diff
@@ -231,7 +231,7 @@ def cli(
     annotate: bool,
     upgrade: bool,
     upgrade_packages: Tuple[str, ...],
-    output_file: Optional[LazyFile],
+    output_file: Union[LazyFile, Union[IO[Any]], None],
```
Comment (webknjaz) [inline piptools/scripts/compile.py:234]: I don't understand why the internal union has just one arg. This seems pointless.

File: piptools/scripts/compile.py [hunk 4]
```diff
@@ -282,7 +282,8 @@ def cli(
 
         # Close the file at the end of the context execution
         assert output_file is not None
-        ctx.call_on_close(safecall(output_file.close_intelligently))
+        if isinstance(output_file, LazyFile):
```
Comment (webknjaz) [inline piptools/scripts/compile.py:285]: Why is this necessary?

**Comment by ssbarnea** [comment]
It is needed to pass mypy because the type changed in click 8 vs 7.

File: piptools/scripts/compile.py (hunk 3 above)
Comment (ssbarnea) [reply piptools/scripts/compile.py]: That was undesired, leftover now removed.

File: piptools/scripts/compile.py (hunk 4 above)
Comment (ssbarnea) [reply piptools/scripts/compile.py]: I added an explanatory comment (click8)

**Review by webknjaz — APPROVED** [review]
